tests/test_smt.py

```diff
@@ -0,0 +1,77 @@
+from hypothesis import (
+        given,
+        strategies as st,
+    )
+
+from trie.smt import (
+    SparseMerkleTree,
+    SparseMerkleProof,
+    calc_root,
+    BLANK_NODE,
+)
+
+
+@given(
+    k=st.binary(min_size=1, max_size=32),
+    v=st.binary(min_size=1, max_size=32),
+)
+def test_simple_kv(k, v):
+    smt = SparseMerkleTree(key_size=len(k))
+
+    # Nothing has been added yet
+    assert not smt.exists(k)
+
+    # Now that something is added, it should be consistent
+    smt.set(k, v)
+    assert smt.get(k) == v
+    assert smt.root_hash == calc_root(k, v, smt.branch(k))
+
+    # If you delete it, it goes away
+    smt.delete(k)
+    assert not smt.exists(k)
+
+
+@given(
+    data=st.data(),
+    key_size=st.integers(min_value=1, max_value=32),
+)
+def test_branch_updates(data, key_size):
+    num_elements = data.draw(st.integers(min_value=1, max_value=key_size*8//10+1))
+    keys = data.draw(
+        st.lists(
+            elements=st.binary(min_size=key_size, max_size=key_size),
+            min_size=1,
+            max_size=num_elements,
+            unique=True,
+        )
+    )
+    vals = data.draw(
+        st.lists(
+            elements=st.binary(min_size=1, max_size=32),
+            min_size=len(keys),
+            max_size=len(keys),
+        )
+    )
+
+    # Empty tree
+    smt = SparseMerkleTree(key_size=key_size)
+
+    # NOTE: smt._get internal method is used for testing only
+    EMPTY_NODE_HASHES = list(smt._get(keys[0])[1])
+
+    # Objects to track proof data
+    proofs = dict([(k, SparseMerkleProof(k, BLANK_NODE, EMPTY_NODE_HASHES)) for k in keys])
+
+    # Track the big list of all updates
+    proof_updates = []
+    for k, p in proofs.items():
+        # Update the key in the smt a bunch of times
+        for v in vals:
+            proof_updates.append((k, v, smt.set(k, v)))
+
+        # Merge all of the updates into the tracked proof entries
+        for u in proof_updates:
+            p.merge(*u)
```

Comment:
`smt.set()` returns what the node updates are for a given change in value in the tree. These updates can be synchronized with a separate proof object (perhaps on a different computer), and that proof object will find the relevant sibling to keep itself up to date so that the proof object can track the same smt root separately without the extra data overhead of managing the whole tree (and the stream of all key-value updates). The storage size of managing the tree can grow non-linearly, whereas the proof object size will remain constant. This is at the expense of increased networking overhead, but that can be optimized by only communicating these updates to the practical depth of the tree (beyond which all keys map to single-branch sub-trees).

My intended use case is mobile, so this tradeoff of storage size is desirable. If my use case had 1m users, the tree would grow to be 2+ gigs in size (without any optimizations).

Reply:
Sorry, this was a comment from before I read the whole thing. Meant to delete it, but it snuck in on my way out the door. Totally on board. :+1: